Use I18n.t string for SHF application state

weedySeaDragon opened the pull request:
### PT Story: admin membership view:  needs I18n
https://www.pivotaltracker.com/story/show/153963489

### Changes proposed in this pull request:
1.  update  SHF application views to use translated state
    - use existing AASM gem method that already handles localiation: `aasm.human_state`
    - write a helper method to return the list for the select list on admin view
2. re-organize locale file entries to use conventions/organization expected by AASM gem
3. update tests so they expect the translated entries
4. rename .feature files from 'membership_application...' to 'shf_application...'
5. refactor .feature file: separate into a feature for showing a single SHF application and a feature for showing a list of SHF applications


### Screenshots (Optional):

<img width="957" alt="admin-applications-sv" src="https://user-images.githubusercontent.com/673794/34655322-5c5fda0c-f3bc-11e7-9753-a7ac6f6c0ceb.png">

<img width="956" alt="admin-applications-en" src="https://user-images.githubusercontent.com/673794/34655321-5c4a04ac-f3bc-11e7-8577-0359ffae8b95.png">

Ready for review:
@AgileVentures/shf-project-team

weedySeaDragon committed:
use AASM activrecord:attributes locale entry convention

weedySeaDragon committed:
helper to list all states with current I18n.locale (do not cache)

weedySeaDragon committed:
restore locale settings after specs

weedySeaDragon committed:
add 'shf applications' to list of paths (to eventually totally replace 'membership applications'

weedySeaDragon committed:
use AASM gem provided ".human_state" method ( = localizes the state name)

weedySeaDragon committed:
use new helper method to display I18n state names

weedySeaDragon committed:
extract from show_membership_application: just the "all..." methods; rename to shf_application

weedySeaDragon committed:
extract from show_membership_application: just the status (state) info, add I18n

weedySeaDragon committed:
status I18n; add note about false positives for sorting

weedySeaDragon committed:
I18n status (state); rename file

weedySeaDragon committed:
refactored files: renamed to shf_application...

weedySeaDragon committed:
update expectations now that the correct I18n statuses are showing up in the search list too

patmbolger commented on features/show_shf_application.feature:
```diff
@@ -0,0 +1,177 @@
+
+
+  @selenium, @member
+  Scenario: Member sees the status translated correctly (default locale, then :en, then :sv)
```
Is this meant to be empty?  Or should be removed?

weedySeaDragon replied:
whoops.  No - shouldn't be empty.

patmbolger commented on features/update_membership_application.feature:
```diff
@@ -1,19 +1,32 @@
+  landing page, there will be 1 time where the status will match the
+  *search* list. So if there are 3 applications in a particular state/status,
+  you should expect to see the status *4* times total (3 times for the applications
+  + 1 time in the list of all statuses/states in the search selection list.
```
We should hide the search form so that the count of search occurrences matches what the _user_ would see.

weedySeaDragon replied:
That _should_ work, except that when  I try that, I still cannot get capybara + selenium + ajax stuff to play well together.

here's what I tried:
1. mark the scenario with @selenium 
2. add these steps at the start (after logging in, going to the right page):
```
    # Hide the search form
    And I click on t("toggle.company_search_form.hide")
    And I wait for all ajax requests to complete
    # verify that it is hidden (otherwise the counts (numbers) will be off below)
    And I should not see t("search")
```

It would fail on the "And I should not see t("search")" step -- the "search" button is still showing up. (it's not really being hidden)

So for now, this passes and is documented well.  
Can we do a different PT chore & PR to address this?

patmbolger replied:
I can get it to work by _not_ including the last step above.

I think the step is finding `Sök` via a match of the (show search form (Swedish)) label, ala `Visa Sökfunktionen`.
```
expected not to find text "Sök" in "Skip to content HALLÅ, ADMIN ALLA INKOMNA ANSÖKNINGAR Visa Sökfunktionen ...
```

weedySeaDragon replied:
aha! thanks for noting that.  It's working by hiding the search form now.  👍

patmbolger commented on spec/helpers/shf_applications_helper_spec.rb:
```diff
@@ -2,6 +2,38 @@
+    end
+
+    it 'correct locale if changed to :en from locale = :sv' do
+      orig_locale = I18n.locale
```
This is handled by the `around(:each)` method.

patmbolger commented on spec/helpers/shf_applications_helper_spec.rb:
```diff
@@ -2,6 +2,38 @@
+      select_list.each do |each_option|
+        expect(each_option[0]).to eq I18n.t("activerecord.attributes.shf_application.state/#{each_option[1]}", locale: :en)
+      end
+      I18n.locale = orig_locale
```
ditto

weedySeaDragon replied:
_it's clear how I arrived at the "gee - I should use `around(...)`!" approach_

patmbolger replied:
Yes, been there done that :)

patmbolger reviewed (changes requested):
Looks great - please see comments.

weedySeaDragon committed:
remove old code for caching original locale

weedySeaDragon committed:
complete scenario for a Member sees status translated; use a better firstname

weedySeaDragon committed:
assign a name to the feature

weedySeaDragon committed:
Merge branch 'develop' into sprint33-i18n-shf-app-states-#153963489

patmbolger reviewed (changes requested):
Updated comment.

weedySeaDragon committed:
Merge branch 'develop' into sprint33-i18n-shf-app-states-#153963489

weedySeaDragon committed:
note sorting TBD question with state translation + ransacker

weedySeaDragon committed:
add step to hide|show search form (frequently used)

weedySeaDragon committed:
no need to wait after pressing the hide/show button

weedySeaDragon committed:
hide the search form so the counts are accurate

weedySeaDragon committed:
Merge branch 'develop' into sprint33-i18n-shf-app-states-#153963489

weedySeaDragon committed:
sort the list to rely on results; name feature; change Co. names to sort better

weedySeaDragon committed:
add step to scroll to the top to avoid ChromeSelenium error when an element is out of view

weedySeaDragon committed:
sort by company number; sort background data to coordinate with expected results

weedySeaDragon committed:
be explicit about what is expected in the view; change date because real date change caused different data displayed

Time passing in the real world meant that the words displayed in the "created date" had changed.  I changed the expectations to allow for that.

weedySeaDragon commented:
I updated the features:
- hide the search form so that _false positives_ don't happen
- sort the lists so that we know exactly what items should appear on which pages
- had to _scroll to the top_ in 1 test because the list of what is displayed is so long that an element is no longer "visible" on the browser.  There's probably a better and more general way to address this in our tests (do something to prevent this from happening at all, perhaps), but this works for now.

weedySeaDragon committed:
Merge branch 'develop' into sprint33-i18n-shf-app-states-#153963489

# Conflicts:
#	features/manage_users.feature

thesuss commented:
Only needs updating to be merged 👍

thesuss commented:
Will make a new PR for this.
